timedctl/timedctl.py

```diff
@@ -25,708 +24,568 @@
+                customer_id = self.get_customer_by_name(
+                    customers, customer_name, archived,
+                )
+            # get the project id
+            projects = self.timed.projects.get(
+                cached=True,
+                filters={"customer": customer_id},
+            )
+            project_id = self.get_project_by_name(
+                projects,
+                project_name,
+                customer_id,
+                archived,
+            )
+        # get the tasks for the specified project
+        tasks = self.timed.tasks.get(cached=True, filters={"project": project_id})
+        output = []
+        for task in tasks:
+            output.append({"id": task["id"], "name": task["attributes"]["name"]})
+        output_formatted(output, output_format)
+
+    def get_overtime(self, date):
+        """Get overtime of user."""
+        user = self.timed.users.me["id"]
+        overtime = self.timed.overtime.get({"user": user, "date": date})
+        msg(f"Current overtime is: {overtime}")
+
+    def get_reports(self, date):
+        """Get reports."""
+        reports = self.timed.reports.get(
+            filters={"date": date},
+            include="task,task.project,task.project.customer",
+        )
+        table = [["Customer", "Project", "Task", "Comment", "Duration"]]
+        for report in reports:
+            task_obj = report["relationships"]["task"]
+            project_obj = task_obj["relationships"]["project"]
+            customer_obj = project_obj["relationships"]["customer"]
+            # get name attributes
+            task, project, customer = (
+                x["attributes"]["name"] for x in [task_obj, project_obj, customer_obj]
+            )
+            comment = report["attributes"]["comment"]
+            duration = report["attributes"]["duration"]
+
+            table.append([customer, project, task, comment, duration])
+        # create the output
+        output = terminaltables.SingleTable(table)
+        msg(f"Reports for {date if date is not None else 'today'}:")
+        click.echo(output.table)
+        msg(f"Total: {time_sum(table)}")
+
+    def get_activities(self, date):
+        """Get activities."""
+        activities = self.timed.activities.get(
+            filters={"day": date},
+            include="task,task.project,task.project.customer",
+        )
+        table = [["Activity", "Comment", "Start", "End"]]
+        for activity_obj in activities:
+            table.append(
+                [
+                    self.format_activity(activity_obj),
+                    activity_obj["attributes"]["comment"],
+                    activity_obj["attributes"]["from-time"].strftime("%H:%M:%S"),
+                    activity_obj["attributes"]["to-time"].strftime("%H:%M:%S")
+                    if activity_obj["attributes"]["to-time"] is not None
+                    else "active",
+                ],
+            )
```

Comment:
Consider using a list comprehension e.g:
```suggestion
table = [
    ["Activity", "Comment", "Start", "End"],
    *[
        [
            self.format_activity(activity_obj),
            activity_obj["attributes"]["comment"],
            activity_obj["attributes"]["from-time"].strftime("%H:%M:%S"),
            activity_obj["attributes"]["to-time"].strftime("%H:%M:%S")
            if activity_obj["attributes"]["to-time"] is not None
            else "active",
        ]
        for activity_obj in activities
    ],
]
```

Reply:
example doesn't work

Reply:
Someone linked this in dev channel and I saw it, so here my 50cc about this.

I would not use a list comprehension and instead make the code more readable, by binding variables. This the minimum I would recommend.

```python
        table = [["Activity", "Comment", "Start", "End"]]
        for activity_obj in activities:
            attributes = activity_obj["attributes"]
            table.append(
                [
                    self.format_activity(activity_obj),
                    attributes["comment"],
                    attributes["from-time"].strftime("%H:%M:%S"),
                    attributes["to-time"].strftime("%H:%M:%S")
                    if attributes["to-time"] is not None
                    else "active",
                ],
            )
 ```
 
  I would probably even do:
  
  ```python
        table = [["Activity", "Comment", "Start", "End"]]
        for activity_obj in activities:
            attributes = activity_obj["attributes"]
            comment = attributes["comment"]
            from_time_fmt = attributes["from-time"].strftime("%H:%M:%S")
            to_time = ttributes["to-time"]
            to_time_fmt = to_time_ftm.strftime("%H:%M:%S")
            table.append(
                [
                    self.format_activity(activity_obj),
                    comment,
                    from_time_fmt,
                    to_time_fmt
                    to_time is not None
                    else "active",
                ],
            )
 ```